go/main.go

```diff
@@ -0,0 +1,164 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"time"
+
+	"github.com/google/go-github/v68/github"
+)
+
+const (
+	SlsaSourceLevel1 = "SLSA_SOURCE_LEVEL_1"
+	SlsaSourceLevel2 = "SLSA_SOURCE_LEVEL_2"
+)
+
+type activity struct {
+	Id           int
+	Before       string
+	After        string
+	Ref          string
+	Timestamp    time.Time
+	ActivityType string `json:"activity_type"`
+}
+
+// Checks to see if the rule meets our requirements.
+func checkRule(ctx context.Context, gh_client *github.Client, owner string, repo string, rule *github.RepositoryRule, minTime time.Time) (bool, error) {
+	ruleset, _, err := gh_client.Repositories.GetRuleset(ctx, owner, repo, rule.RulesetID, false)
+	if err != nil {
+		return false, err
+	}
+
+	// We need rules to be 'active' and to have been updated no later than minTime.
+	if ruleset.Enforcement != "active" {
+		return false, nil
+	}
+
+	if minTime.Before(ruleset.UpdatedAt.Time) {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func commitPushTime(ctx context.Context, gh_client *github.Client, commit string, owner string, repo string, branch string) (time.Time, error) {
+	// Unfortunately the gh_client doesn't have native support for this...'
```

Comment:
we should probably enhance this comment. 

```
we need to find a trustworthy timestamp (and we don't trust the commit metadata).
On GitHub with simple repos that require use of a squash merge strategy, each PR will add exactly one commit to the target branch, that commit will be _novel_ (meaning it was produced by the PR application) and it will be the after_sha of the ref update.

In that situation, you can use the /activity endpoint to get the timestamp when GitHub server received the ref-update.

Reply:
Yeah definitely.

Separately, maybe this tool could require squash merge if that makes things easier?

I think it would be fine for the tool to be more opinionated than the SLSA spec itself.